client/mm/exchange_adaptor.go

```diff
@@ -2897,8 +2907,78 @@ func (u *unifiedExchangeAdaptor) optimizeTransfers(dist *distribution, dexSellLo
 	quoteInv.toWithdraw = split.quoteWithdraw
 }
 
-// transfer attempts to perform the transers specified in the distribution.
-func (u *unifiedExchangeAdaptor) transfer(dist *distribution, currEpoch uint64) (actionTaken bool, err error) {
+// tryInternalTransfers attempts to use available funds that are not reserved
+// by any bots to rebalance between the DEX and the CEX instead of doing an
+// actual deposit or withdrawal. True is returned if the deposits and/or
+// withdrawals were completely covered using available funds.
+func (u *unifiedExchangeAdaptor) tryInternalTransfers(dist *distribution) bool {
+	doTransfer := func(dexAvail, cexAvail map[uint32]uint64) bool {
+		u.balancesMtx.Lock()
+		defer u.balancesMtx.Unlock()
+
+		baseInv, quoteInv := dist.baseInv, dist.quoteInv
+		dexDiffs, cexDiffs := make(map[uint32]int64), make(map[uint32]int64)
+
+		complete := true
+		transferDone := false
+
+		if baseInv.toDeposit > 0 {
+			botDEXBal := u.dexBalance(u.baseID).Available
+			toDeposit := utils.Min(baseInv.toDeposit, cexAvail[u.baseID], botDEXBal)
+			complete = complete && toDeposit == baseInv.toDeposit
+			transferDone = transferDone || toDeposit > 0
+			u.baseCexBalances[u.baseID] += int64(toDeposit)
+			u.baseDexBalances[u.baseID] -= int64(toDeposit)
+			dexDiffs = map[uint32]int64{u.baseID: -int64(toDeposit)}
+			cexDiffs = map[uint32]int64{u.baseID: int64(toDeposit)}
+		}
+
+		if baseInv.toWithdraw > 0 {
+			botCEXBal := u.cexBalance(u.baseID).Available
+			toWithdraw := utils.Min(baseInv.toWithdraw, dexAvail[u.baseID], botCEXBal)
+			complete = complete && toWithdraw == baseInv.toWithdraw
+			transferDone = transferDone || toWithdraw > 0
+			u.baseCexBalances[u.baseID] -= int64(toWithdraw)
+			u.baseDexBalances[u.baseID] += int64(toWithdraw)
+			dexDiffs = map[uint32]int64{u.baseID: int64(toWithdraw)}
+			cexDiffs = map[uint32]int64{u.baseID: -int64(toWithdraw)}
+		}
+
+		if quoteInv.toDeposit > 0 {
+			botDEXBal := u.dexBalance(u.quoteID).Available
+			toDeposit := utils.Min(quoteInv.toDeposit, cexAvail[u.quoteID], botDEXBal)
+			complete = complete && toDeposit == quoteInv.toDeposit
+			transferDone = transferDone || toDeposit > 0
+			u.baseCexBalances[u.quoteID] += int64(toDeposit)
+			u.baseDexBalances[u.quoteID] -= int64(toDeposit)
+			dexDiffs[u.quoteID] = -int64(toDeposit)
+			cexDiffs[u.quoteID] = int64(toDeposit)
+		}
+
+		if quoteInv.toWithdraw > 0 {
+			botCEXBal := u.cexBalance(u.quoteID).Available
+			toWithdraw := utils.Min(quoteInv.toWithdraw, dexAvail[u.quoteID], botCEXBal)
+			complete = complete && toWithdraw == quoteInv.toWithdraw
+			transferDone = transferDone || toWithdraw > 0
+			u.baseCexBalances[u.quoteID] -= int64(toWithdraw)
+			u.baseDexBalances[u.quoteID] += int64(toWithdraw)
+			dexDiffs[u.quoteID] = int64(toWithdraw)
+			cexDiffs[u.quoteID] = -int64(toWithdraw)
+		}
+
+		if transferDone {
+			u.logBalanceAdjustments(dexDiffs, cexDiffs, "Internal transfer")
+		}
+
+		return complete
```

Comment:
If a partial internal transfer doesn't improve our matchability, is it worth doing? In some since, we know that it probably won't improve our matchability as much as an external transfer, since `optimizeTransfers` already optimized the transfer amounts. And a partial internal transfer risks putting our optimized transfer amounts lower than our min transfer settings, resulting in a lower matchability state in the end.

Reply:
You're right, I'm doing a bit of a refactor.